lib/src/barback/ddc_transformer.dart

```diff
@@ -0,0 +1,500 @@
+      : new File(p.joinAll([rootDir]..addAll(p.url.split(id.path))));
+}
+
+Set<AssetId> _findSummaryIds(package) {
+  // TODO(jakemac): Read build.yaml if available?
+  return new Set<AssetId>()
+    ..add(
+        new AssetId(package, p.url.join('lib', '$package.$_summaryExtension')));
+}
+
+Future<bool> _isEntryPoint(
+    AssetId id, TransformLogger logger, _InputAsStringReader readInputAsString,
+    {Set<AssetId> seenIds}) async {
+  seenIds ??= new Set<AssetId>();
+  if (!seenIds.add(id)) return false;
+
+  var content = await readInputAsString(id);
+  // Suppress errors at this step, dartdevc will error later if the file can't
+  // parse.
+  var unit = parseCompilationUnit(content, name: '$id', suppressErrors: true);
+  // Logic to check for a valid main(). Currently this is a check for any
+  // top level function called "main" with 2 or fewer parameters. The 2nd
+  // argument is allowed for the isolate conversion test.
+  if (unit.declarations.any((d) =>
+      d is FunctionDeclaration &&
+      d.name.name == 'main' &&
+      d.functionExpression.parameters.parameters.length <= 2)) {
+    return true;
+  }
+
+  // Additionally search all exports that might be exposing a `main` based on
+  // the show/hide settings.
+  var exportsToSearch = unit.directives.where((d) {
+    if (d is! ExportDirective) return false;
+    for (var combinator in (d as ExportDirective).combinators) {
+      if ('${combinator.keyword}' == 'show') {
+        if (combinator.childEntities.any((e) => '$e' == 'main')) return true;
+      } else if ('${combinator.keyword}' == 'hide') {
+        if (combinator.childEntities.any((e) => '$e' == 'main')) return false;
+      }
+    }
+    return true;
+  });
+
+  for (var export in exportsToSearch) {
+    var exportId = _urlToAssetId(
+        id, (export as UriBasedDirective).uri.stringValue, logger);
+    if (exportId == null) continue;
+    if (await _isEntryPoint(exportId, logger, readInputAsString,
+        seenIds: seenIds)) return true;
+  }
+  return false;
+}
+
+AssetId _urlToAssetId(AssetId source, String url, TransformLogger logger) {
+  var uri = Uri.parse(url);
+  if (uri.isAbsolute) {
+    if (uri.scheme == 'package') {
+      var parts = uri.pathSegments;
+      return new AssetId(
+          parts.first, p.url.joinAll(['lib']..addAll(parts.skip(1))));
+    } else if (uri.scheme == 'dart') {
+      return null;
+    } else {
+      logger.error('Unable to resolve import. Only package: paths and relative '
+          'paths are supported, got `$url`');
+      return null;
+    }
+  } else {
+    // Relative path.
+    var targetPath =
+        p.url.normalize(p.url.join(p.url.dirname(source.path), uri.path));
+    return new AssetId(source.package, targetPath);
+  }
+}
+
+Future _writeFile(File file, Stream<List<int>> stream) async {
+  await file.create(recursive: true);
+  var sink = file.openWrite();
+  await sink.addStream(stream);
+  await sink.close();
+}
+
+const _summaryExtension = 'api.ds';
```

Comment:
Nit: I think we generally keep top level variables and constants at the top of files.